boot/bootutil/src/encrypted_psa.c

```diff
@@ -0,0 +1,454 @@
+/*
+ * Copyright (c) 2025 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "mcuboot_config/mcuboot_config.h"
+
+#include <stddef.h>
+#include <inttypes.h>
+#include <string.h>
+
+/* We are not really using the MBEDTLS but need the ASN.1 parsing functions */
+#define MBEDTLS_ASN1_PARSE_C
+
+#include "bootutil/crypto/sha.h"
+#include "mbedtls/oid.h"
+#include "mbedtls/asn1.h"
+
+#include "bootutil/image.h"
+#include "bootutil/enc_key.h"
+#include "bootutil/sign_key.h"
+#include "bootutil/crypto/common.h"
+
+#include "bootutil_priv.h"
+#include "bootutil/bootutil_log.h"
+
+BOOT_LOG_MODULE_DECLARE(mcuboot_psa_enc);
+
+#define EXPECTED_ENC_LEN    BOOT_ENC_TLV_SIZE
+#define EXPECTED_ENC_TLV    IMAGE_TLV_ENC_X25519
+#define EC_PUBK_INDEX       (0)
+#define EC_TAG_INDEX        (32)
+#define EC_CIPHERKEY_INDEX  (32 + 32)
+_Static_assert(EC_CIPHERKEY_INDEX + BOOT_ENC_KEY_SIZE == EXPECTED_ENC_LEN,
+        "Please fix ECIES-X25519 component indexes");
+
+#define X25519_OID "\x6e"
+static const uint8_t ec_pubkey_oid[] = MBEDTLS_OID_ISO_IDENTIFIED_ORG \
+                                       MBEDTLS_OID_ORG_GOV X25519_OID;
+
+#define SHARED_KEY_LEN 32
+#define PRIV_KEY_LEN   32
+
+/* Fixme: This duplicates code from encrypted.c and depends on mbedtls */
+static int
+parse_x25519_enckey(uint8_t **p, uint8_t *end, uint8_t *private_key)
+{
+    size_t len;
+    int version;
+    mbedtls_asn1_buf alg;
+    mbedtls_asn1_buf param;
+
+    if (mbedtls_asn1_get_tag(p, end, &len, MBEDTLS_ASN1_CONSTRUCTED |
+                                           MBEDTLS_ASN1_SEQUENCE) != 0) {
+        return -1;
```

Comment:
should have defines for these magic numbers

Reply:
Yeah, this is the same as for original code in other ASN parsers here.
This should be probably universally replaced with some one FAIL value.